### PR TITLE
Add OreRegistry for unifying basic materials

### PR DESCRIPTION
**Pull request by mezz**
This aims to slice off a piece of functionality from `OreDictionary` and make it something better.
Currently the `OreDictionary` seems to do a few different things all together, and it's generally disliked and needs replacement. This is one step towards that.

@Nedelosk and I have worked on a mod to brainstorm how to do this already (it is also called OreRegistry), but really this system needs to be in Forge to be effective at all. I used that mod as a starting point for this PR. I was able to simplify it a lot by being able to change things in Forge like recipes, and 1.12's ingredients and JSON recipes have made things much simpler as well.

Since this is a relatively big feature, it needs community feedback.
Below I'll describe the system, taken from the javadoc on `OreRegistry.java`.
Before giving feedback please see the code as well.

**Description:**

The Ore Registry allows mods to register *basic* material items so they can be unified.
Using the Ore Registry, everything can return a single unified type, so for example players will only ever get one kind of Copper Ingot ItemStack.

Basic materials follow vanilla conventions, like iron ore being smeltable into iron ingots.
If other mods cannot make basic assumptions about your materials, do not register them with the OreRegistry.

Recipes created with the Ore Registry can accept any item registered for a specific type, and return just the one unified type.
Blocks are not unified, but ores generated in the world should be written to drop the unified ore ItemStack. True ore block worldgen unification for modpacks is completely out of scope of this system, and has existing solutions.

**Usage:**

First, create your `OreMaterial`s and `OreShape`s or use the ones already created in `OreRegistryConstants`.

Then, add your ItemStack with `OreRegistry#addOre(OreMaterial, OreShape, ItemStack)`.
You will get back an instance of `BasicOre`, which is used to unify all ores with the specified Material and Shape.

**Recipes:**

` BasicOre` can be used in recipes like `ShapedForgeRecipe` and `ShapelessForgeRecipe` (which have been written to replace the old Forge ore recipes) or you can implement your own recipes using them.

You can also create JSON recipes that use the OreRegistry for ingredients or results. 
To have OreRegistry results, use the recipe type `"forge:shaped"` or `"forge:shapeless"`.
To have OreRegistry ingredients, use the ingredient format:
`{"type":"forge:ore_registry", "material": "materialUid", "shape": "shapeUid"}`
For examples see the JSON recipes in the forge test directory `src\test\resources\assets\ore_registry_test\recipes`.

### CODE REVIEW

**Comment by KnightMiner** [comment]
I was hoping something like this would be added, using the oredictionary for the sake of unification of outputs has lead to many issues in the past.

How does this system work with entries that are empty? For example, if I add integration recipes for making any ingot into its dust variant, but I do not register any dusts myself? I don't see anything in the code directly referencing this, but the easiest solution I can think of while still ensuring the recipe is registered in case the entry changes later is to check if the output exists (is not empty) before returning true in the matches method.

**Comment by mezz** [comment]
Currently there are no empty entries. I'm not sure the best way to handle it. OreDictionary handles empty entries but it's also a point of confusion for modders. On the other hand it would simplify things a bit, like the case you described.

**Comment by KingLemming** [comment]
Yeah, empty entries are a thing that the oredict maybe shouldn't support, tbh, but I know that's an argument that we've had and no minds will be changing.

File: src/main/java/net/minecraftforge/common/crafting/CraftingHelper.java [hunk 1]
```diff
@@ -201,6 +211,40 @@ public static Ingredient getIngredient(JsonElement json, JsonContext context)
         return factory.parse(context, obj);
     }
 
+    public static Supplier<ItemStack> getItemStackSupplier(JsonObject json, JsonContext context)
+    {
+        BasicOre basicOre = getBasicOre(json, context);
+        if (basicOre != null)
+        {
+            final int count = JsonUtils.getInt(json, "count", 1);
+            return () ->
```
Comment (Alpvax) [inline src/main/java/net/minecraftforge/common/crafting/CraftingHelper.java]: Is this lambda required?

**Comment by Ivorforce** [comment]
It's probably best not to handle 'empty' entries, but still offer the possibility for a mod to query the multiplicity of a unified 'ore'. If need be, it can handle uncraftable recipes itself that way.

Speaking of which, I'd like a rename of the system. It may have been implemented for ores but obviously it's in use for many other items too. Naming it 'Ore'dictionary specifically is a bit misleading. Since Blocks are out of scope, perhaps simply 'Item'Dictionary (or as PaleoCrafter suggested, ItemThesaurus) would be fitting.

**Comment by Alpvax** [comment]
I would also prefer to rename OreShape to something more meaningful, When I first read this, I had to read through the code to work out exactly what it was. Something like OreType or MaterialType or ItemType or something along those lines. I don't think of "gem" "dust" and "ingot" as 'shapes'.

@liach I noticed you downvoted. Anything in particular that you are against?

**Comment by liach** [comment]
First, I think this is a lot of code without a lot of meaningful additions. It simply adds some ore priority system. The access of things need a lot of null checks, too.
Second, I fear this affects the usability of ore dict. It is meant to be versatile, like how Pam's mod uses it. 
Third, we talk about why we oredict. I believe ore dicts are traits of an item stack: they are not the only definition. Assume if nether copper is added and is set to drop prioritized copper ore in the ore registry. This situation frequently means that nether copper ores simply drop normal copper ores, which is a bug.

**Comment by RANKSHANK** [comment]
@Alpvax Perhaps MaterialForm?

Anyways that said this is the sort of thing what I was aiming for with #3171 , but I got caught up with the more quantifying/ chemical aspects of the materials as opposed to this more practical approach.

File: src/main/java/net/minecraftforge/oreregistry/OreRegistryConstants.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    public static final OreMaterial SILVER = OreRegistry.addMaterial("silver");
+    public static final OreMaterial STEEL = OreRegistry.addMaterial("steel");
+    public static final OreMaterial TIN = OreRegistry.addMaterial("tin");
+    public static final OreMaterial ZINC = OreRegistry.addMaterial("zinc");
+
+    /* Common Shapes */
+    public static final OreShape ORE = OreRegistry.addShape("ore");
+    public static final OreShape INGOT = OreRegistry.addShape("ingot");
+    public static final OreShape DUST = OreRegistry.addShape("dust");
+    public static final OreShape POWDER = OreRegistry.addShape("powder");
```
Comment (RANKSHANK) [inline src/main/java/net/minecraftforge/oreregistry/OreRegistryConstants.java:58]: What's the difference in application between dust and powder? Or gem and crystal? I think forcing 'shaping conventions' early on is a good idea here

File: src/main/java/net/minecraftforge/oreregistry/BasicOre.java [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+ *
+ * To get an instance, you can add your ore with {@link OreRegistry#addOre(OreMaterial, OreShape, ItemStack)}.
+ * If you have no ore you can just retrieve one with {@link OreRegistry#getOre(OreMaterial, OreShape)}.
+ *
+ * One variant is selected to be "the one" which is always output from recipes and block drops. See {@link #get()}.
+ * The variant that gets selected can be manipulated by the Forge Config, see "forge.configgui.oreRegistryPriorityModIds"
+ */
+public final class BasicOre implements Supplier<ItemStack>
+{
+    @Nonnull
```
Comment (liach) [inline src/main/java/net/minecraftforge/oreregistry/BasicOre.java:47]: Move these to the package

File: src/main/java/net/minecraftforge/oreregistry/OreRegistryConstants.java (hunk 2 above)
Comment (Alpvax) [reply src/main/java/net/minecraftforge/oreregistry/OreRegistryConstants.java:58]: I believe it is the difference between redstone/glowstone "dust" and ground, smeltable "powders". The question really is whether we need that distinction.

**Comment by Alpvax** [comment]
> Assume if nether copper is added and is set to drop prioritized copper ore in the ore registry. This situation frequently means that nether copper ores simply drop normal copper ores, which is a bug.

This is a very good point. Maybe there is some good, standardised way to implement this. Something like submaterials (e.g. NETHER or alternatively, the material the ore is in, i.e. stone, netherrack, endstone).

It is important to not that this is not a full replacement for the OreDictionary, but this will simplify the "n mods all add copper ore, with n different processing machines and n different ingots, nuggets and dusts"

For some things, the OreDictionary will still be required, as they are not identically functional items (e.g. AE quartz)

**Comment by JamiesWhiteShirt** [comment]
Let's get this out of the way first: Can we deprecate the "Ore" name? The purpose of this system is ItemStack unification, which is why it should be called an ItemStack unification registry, not an ore registry.

I also see potential for generifying this. It could be reused for unification of other things, such as blockstates and fluids.

What should happen when registrations are added in a way that changes the selected item? This will happen if mod IDs are not added to the `oreRegistryPriorityModIds` configuration. Let's say I have a world where silver ingots from Thermal Foundation is king. Then I add Embers, now silver ingots from Embers is king. Seeing as forge already writes persistent data, maybe this order could be persisted?

If a connected client and server are configured differently, OreSortOrder could cause unexpected desync behavior. Should this be considered a user error, or should it be addressed?

**Comment by mezz** [comment]
I'm fine with renaming things, that makes sense.

@liach 

- Your concerns are good, but I am avoiding that functionality here. OreDictionary does a lot, this is not a full replacement and it is not meant to be one.
- The access null checks are a result of not allowing empty entries. This can be changed if it is an issue, but is a compromise between two things.
- For the nether ores example, they'd just be a new "shape", "nether_ore". They are not the same as "ore" but they have a specific relation to it just like ingots or any other shape. For example in their recipe they can make sure they smelt into one or two of the corresponding "ore" shape.

@JamiesWhiteShirt

- This could be generic but it would become a bit bloated. Handling ItemStack properly has a lot of special cases already, I tried making it generic and ended up with a lot of code just checking validity. You can see some of that left over in `RecipeUtil.CheckedItemStackSupplier`.
- I am not sure what the best way to persist this data is, it sounds like a good idea though.
- Server -> Client sync is probably necessary but I'm not sure what the best approach is.

**Comment by Alpvax** [comment]
Has this now been abandoned?
I still think this is quite a good idea.

**Comment by mezz** [comment]
It's not really abandoned, I just need to find some free time to focus on it.
I've been working on a lot of other stuff lately.

**Comment by Alpvax** [comment]
Ah, ok, sorry to pester you then

**Comment by mezz** [comment]
Superseded by https://github.com/MinecraftForge/MinecraftForge/pull/4457

**Pull request closed without merging**